http-client/Data/KeyedPool.hs

```diff
@@ -0,0 +1,300 @@
+--
+-- * A reaper thread will destroy unused idle resources regularly. It
+-- will stop running once 'kpVar' contains a 'PoolClosed' value.
+--
+-- * 'takeKeyedPool' is async exception safe, but relies on the
+-- /caller/ to ensure prompt cleanup. See its comment for more
+-- information.
+module Data.KeyedPool
+    ( KeyedPool
+    , createKeyedPool
+    , takeKeyedPool
+    , Managed
+    , managedResource
+    , managedReused
+    , managedRelease
+    , Reuse (..)
+    , dummyManaged
+    ) where
+
+import Control.Concurrent (forkIO, threadDelay)
+import Control.Concurrent.STM
+import Control.Exception (mask_, catch, SomeException)
+import Control.Monad (join, unless)
+import Data.Map (Map)
+import Data.Maybe (isJust)
+import qualified Data.Map.Strict as Map
+import Data.Time (UTCTime, getCurrentTime, addUTCTime)
+import Data.IORef (IORef, newIORef, mkWeakIORef)
+import qualified Data.Foldable as F
+import GHC.Conc (unsafeIOToSTM)
+import System.IO.Unsafe (unsafePerformIO)
+
+data KeyedPool key resource = KeyedPool
+    { kpCreate :: !(key -> IO resource)
+    , kpDestroy :: !(resource -> IO ())
+    , kpMaxPerKey :: !Int
+    , kpMaxTotal :: !Int
+    , kpVar :: !(TVar (PoolMap key resource))
+    , kpAlive :: !(IORef ())
+    }
+
+data PoolMap key resource
+    = PoolClosed
+    | PoolOpen
+        -- Total number of resources in the pool
+        {-# UNPACK #-} !Int
+        !(Map key (PoolList resource))
+    deriving F.Foldable
+
+-- | A non-empty list which keeps track of its own length and when
+-- each resource was created.
+data PoolList a
+    = One a {-# UNPACK #-} !UTCTime
+    | Cons
+        a
+
+        -- size of the list from this point and on
+        {-# UNPACK #-} !Int
+
+        {-# UNPACK #-} !UTCTime
+        !(PoolList a)
+    deriving F.Foldable
+
+plistToList :: PoolList a -> [(UTCTime, a)]
+plistToList (One a t) = [(t, a)]
+plistToList (Cons a _ t plist) = (t, a) : plistToList plist
+
+plistFromList :: [(UTCTime, a)] -> Maybe (PoolList a)
+plistFromList [] = Nothing
+plistFromList [(t, a)] = Just (One a t)
+plistFromList xs =
+    Just . snd . go $ xs
+  where
+    go [] = error "plistFromList.go []"
+    go [(t, a)] = (2, One a t)
+    go ((t, a):rest) =
+        let (i, rest') = go rest
+            i' = i + 1
+         in i' `seq` (i', Cons a i t rest')
+
+-- | Create a new 'KeyedPool' which will automatically clean up after
+-- itself when all referenced to the 'KeyedPool' are gone. It will
+-- also fork a reaper thread to regularly kill off unused resource.
+createKeyedPool
+    :: Ord key
+    => (key -> IO resource) -- ^ create a new resource
+    -> (resource -> IO ()) -- ^ destroy a resource
+    -> Int -- ^ number of resources per key to allow in the pool
+    -> Int -- ^ number of resources to allow in the pool across all keys
+    -> IO (KeyedPool key resource)
+createKeyedPool create destroy maxPerKey maxTotal = do
+    var <- newTVarIO $ PoolOpen 0 Map.empty
+    _ <- forkIO $ reap destroy var
```

Comment:
I just rearranged the code instead of `bracketOnException`, with a comment justifying why, I think this is correct.